src/renderers/shared/fiber/ReactFiberReconciler.js

```diff
@@ -114,29 +133,62 @@ module.exports = function<T, P, I>(config : HostConfig<T, P, I>) : Reconciler {
       // progress.
       const current = workInProgress.alternate;
       const next = completeWork(current, workInProgress);
+
+      // The work is now done. We don't need this anymore. This flags
+      // to the system not to redo any work here.
+      workInProgress.pendingProps = null;
```

Comment:
I'm not sure about the GC, because once you write the reference once it is scheduled for promotion. I'm not sure if it gets unscheduled for free. Regardless, we could probably find a clever way to do that on the stack if needed.

Currently it all gets moved to memoizedProps by the time completeWork is called so for anything memoized we need to store it until the tree is complete and indefinitely.